proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+        &rule_struct_name,
+        &rule_instance_struct_name,
+        &rule_instance_state_fields,
+    );
+
+    let rule_instance_per_file_struct_name = format_ident!("{}PerFile", rule_instance_struct_name);
+
+    let rule_instance_per_file_state_fields = rule.state.as_ref().map_or_else(
+        || Default::default(),
+        |state| {
+            state
+                .scope_sections
+                .iter()
+                .filter(|scope_section| scope_section.scope == RuleStateScope::PerFileRun)
+                .flat_map(|scope_section| scope_section.fields.iter())
+                .collect::<Vec<_>>()
+        },
+    );
+
+    let rule_instance_rule_instance_impl = get_rule_instance_rule_instance_impl(
+        &rule_instance_struct_name,
+        &rule_instance_per_file_struct_name,
+        &rule_instance_per_file_state_fields,
+    );
+
+    let rule_instance_per_file_struct_definition = get_rule_instance_per_file_struct_definition(
+        &rule_instance_struct_name,
+        &rule_instance_per_file_struct_name,
+        &rule_instance_per_file_state_fields,
+    );
+
+    let rule_instance_per_file_rule_instance_per_file_impl =
+        get_rule_instance_per_file_rule_instance_per_file_impl(
+            &rule,
+            &rule_instance_per_file_struct_name,
+        );
+
+    let instantiate_rule = get_rule_struct_creation(&rule_struct_name, &rule_state_fields);
+
+    quote! {
+        {
+            #rule_struct_definition
+
+            #rule_rule_impl
+
+            #rule_instance_struct_definition
+
+            #rule_instance_rule_instance_impl
+
+            #rule_instance_per_file_struct_definition
+
+            #rule_instance_per_file_rule_instance_per_file_impl
+
+            #instantiate_rule
+        }
+    }
+    .into()
+}
+
+fn get_rule_struct_definition(
+    rule_struct_name: &Ident,
+    rule_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let field_names = rule_state_fields.iter().map(|field| &field.name);
+    let field_types = rule_state_fields.iter().map(|field| &field.type_);
+    quote! {
+        struct #rule_struct_name {
+            #(#field_names: #field_types),*
+        }
+    }
+}
+
+fn get_rule_rule_impl(
+    rule: &Rule,
+    rule_struct_name: &Ident,
+    rule_instance_struct_name: &Ident,
+    rule_instance_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let name = &rule.name;
+    let fixable = match rule.fixable.as_ref() {
+        Some(fixable) => quote!(#fixable),
+        None => quote!(false),
+    };
+    let rule_instance_state_field_names =
+        rule_instance_state_fields.iter().map(|field| &field.name);
+    let rule_instance_state_field_initializers =
+        rule_instance_state_fields
+            .iter()
+            .map(|field| match field.initializer.as_ref() {
+                Some(initializer) => quote!(#initializer),
+                None => quote!(Default::default()),
```

Comment:
Allow omitting the "state" field initializers if they should be initialized to `Default::default()`

Reply:
Could this be done as an `unwrap_or_default`? (Before the `quote!`, I guess?)

Reply:
I _think_ that's conflating "macro-time expansion" vs "compile-time", in this scenario basically "there's nothing there to `unwrap_or_default`" because we're supplying an initialization value (of `Default::default()`) for a state field that they omitted an initialization value for (in the macro contents)